go/p2p/rpc/client.go

```diff
@@ -168,101 +157,168 @@ type AggregateFunc func(rsp interface{}, pf PeerFeedback) bool
 
 // CallMultiOptions are per-multicall options
 type CallMultiOptions struct {
-	aggregateFn AggregateFunc
+	maxPeerResponseTime time.Duration
+	maxParallelRequests uint
+	aggregateFn         AggregateFunc
+}
+
+// NewCallMultiOptions creates options using default and given values.
+func NewCallMultiOptions(opts ...CallMultiOption) *CallMultiOptions {
+	co := CallMultiOptions{
+		maxPeerResponseTime: RequestReadDeadline,
+		maxParallelRequests: DefaultParallelRequests,
+	}
+	for _, opt := range opts {
+		opt(&co)
+	}
+	return &co
 }
 
 // CallMultiOption is a per-multicall option setter.
 type CallMultiOption func(opts *CallMultiOptions)
 
+// WithMaxPeerResponseTimeMulti configures the maximum response time for a peer.
+func WithMaxPeerResponseTimeMulti(d time.Duration) CallMultiOption {
+	return func(opts *CallMultiOptions) {
+		opts.maxPeerResponseTime = d
+	}
+}
+
+// WithMaxParallelRequests configures the maximum number of parallel requests to make.
+func WithMaxParallelRequests(n uint) CallMultiOption {
+	return func(opts *CallMultiOptions) {
+		opts.maxParallelRequests = n
+	}
+}
+
 // WithAggregateFn configures the response aggregation function to use.
 func WithAggregateFn(fn AggregateFunc) CallMultiOption {
 	return func(opts *CallMultiOptions) {
 		opts.aggregateFn = fn
 	}
 }
 
+// ClientListener is an interface for an object wishing to receive notifications from the client.
+type ClientListener interface {
+	// RecordSuccess is called on a successful protocol interaction with a peer.
+	RecordSuccess(peerID core.PeerID, latency time.Duration)
+
+	// RecordFailure is called on an unsuccessful protocol interaction with a peer.
+	RecordFailure(peerID core.PeerID, latency time.Duration)
+
+	// RecordBadPeer is called when a malicious protocol interaction with a peer is detected.
+	RecordBadPeer(peerID core.PeerID)
+}
+
 // Client is an RPC client for a given protocol.
 type Client interface {
-	PeerManager
-
-	// Call attempts to route the given RPC method call to one of the peers that supports the
-	// protocol based on past experience with the peers.
+	// Call attempts to route the given RPC method call to the given peer. It's up to the caller
+	// to provide only connected peers that support the protocol.
 	//
 	// On success it returns a PeerFeedback instance that should be used by the caller to provide
 	// deferred feedback on whether the peer is any good or not. This will help guide later choices
 	// when routing calls.
 	Call(
```

Comment:
What is the difference between `Call` and `CallOne` with only a single peer specified? Should we just get rid of this one?

Reply:
I wanted to get rid of `validationFn` and didn't want to have in the logs that calls to all peers failed. But now that I look back it is better to have only one method. 
Actually, I still prefer having two methods, I will just change that `Call` calls `CallOne`.